corehq/util/metrics/datadog.py

```diff
@@ -12,12 +13,17 @@
 datadog_logger = logging.getLogger('datadog')
 
 
+def _sanitize_value(tag_value: str):
+    # valid characters for tag values from https://docs.datadoghq.com/getting_started/tagging/#defining-tags
+    return re.sub(r'[^a-zA-Z0-9_:./-]+', '_', tag_value.lower())
+
+
 def _format_tags(tag_values: Dict[str, str]):
     if not tag_values:
         return None
 
     return [
-        f'{name}:{value}' for name, value in tag_values.items()
+        f'{name}:{_sanitize_value(value)}' for name, value in tag_values.items()
     ]
```

Comment:
Can you talk a bit about the choice to do it here (applying to all datadog metrics we send) versus in the one place this came up? This is more comprehensive but magical, that would be more explicit but won't cover other places it might be happening.

Reply:
Yeah, since these metrics otherwise fail, there didn't seem to be a benefit to letting them fail, and rather than having everyone rediscover and reimplement these rules at each call site, it seemed much easier to just do it at submission. Furthermore, if the rules ever change, there is only one place to update. What is the advantage of the other version be?

Reply:
I guess I'm just a little uncomfortable quietly transforming from a larger string space to a smaller string space, when really it's a sign that the caller doesn't understand the string space constraint their working in. If we're passing in strings with spaces, those strings are probably coming straight from users. If we then transform them, there's the chance that users can enter two different strings that are mapped onto the same tag value. Applying that everywhere quietly seems like it's hiding potential errors.

I think my favorite local fix would be to find a unique string that identifies accounts that is also fits the string space datadog accepts, but that may not be possible. It would be nice if there were a canonical way to map a unicode string onto a constrained space in a way where most of the characters stayed the same but unavailable characters were replaced with an escape sequence.

But maybe I'm going overboard here. I would insist on it if it were being used to identify something in a database, as that would create an obvious horizontal escalation vulnerability (user craft string that gets mapped to a string that can be used to look up other user's data), which is not the case here.